src/Symfony/Component/ArgumentResolver/ArgumentResolver.php

```diff
@@ -0,0 +1,145 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\ArgumentResolver;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\ArgumentResolver\ArgumentMetadata\ArgumentMetadata;
+use Symfony\Component\ArgumentResolver\ArgumentMetadata\ArgumentMetadataFactory;
+use Symfony\Component\ArgumentResolver\ArgumentMetadata\ArgumentMetadataFactoryInterface;
+use Symfony\Component\ArgumentResolver\Exception\InvalidArgumentException;
+use Symfony\Component\ArgumentResolver\Exception\LogicException;
+use Symfony\Component\ArgumentResolver\Exception\NearMissValueResolverException;
+use Symfony\Component\ArgumentResolver\Exception\RuntimeException;
+use Symfony\Component\ArgumentResolver\ValueResolver\DefaultValueResolver;
+use Symfony\Component\ArgumentResolver\Attribute\ValueResolver;
+use Symfony\Component\ArgumentResolver\Exception\ResolverNotFoundException;
+use Symfony\Contracts\Service\ServiceProviderInterface;
+use Symfony\Component\ArgumentResolver\ValueResolver\ValueResolverInterface;
+
+/**
+ * Responsible for resolving the arguments passed to an action.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ * @author Robin Chalas <robin@baksla.sh>
+ */
+class ArgumentResolver implements ArgumentResolverInterface
+{
+    private readonly ArgumentMetadataFactoryInterface $argumentMetadataFactory;
+
+    /**
+     * @param iterable<mixed, ValueResolverInterface> $argumentValueResolvers
+     */
+    public function __construct(
+        ?ArgumentMetadataFactoryInterface $argumentMetadataFactory = null,
+        private iterable $argumentValueResolvers = [],
+        private readonly ?ContainerInterface $namedResolvers = null,
+    ) {
+        $this->argumentMetadataFactory = $argumentMetadataFactory ?? new ArgumentMetadataFactory();
+        $this->argumentValueResolvers = $argumentValueResolvers ?: static::getDefaultValueResolvers();
+    }
+
+    public function getArguments(mixed $input, callable $callable, ?\ReflectionFunctionAbstract $reflector = null): array
```

Comment:
```suggestion
    final public function getArguments(mixed $input, callable $callable, ?\ReflectionFunctionAbstract $reflector = null): array
```
If don't think we want this to be overridden.

For the `ControllerArgumentResolver::getArguments`, maybe we can create a new `abstract protected getSubject` (or any other name) method to fit with a template-method-like design pattern?